lib/autoproj/daemon/pull_request_cache.rb

```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
+
+require 'autoproj'
+require 'autoproj/daemon/github/pull_request'
+require 'yaml'
+
+module Autoproj
+    # Main plugin module
+    module Daemon
+        # A PullRequest cache
+        class PullRequestCache
+            # @return [Array<CachedPullRequest>]
+            attr_reader :pull_requests
+
+            # @return [Autoproj::Workspace]
+            attr_reader :ws
+
+            def initialize(workspace)
+                @pull_requests = []
+                @ws = workspace
+            end
+
+            CachedPullRequest = Struct.new :base_owner, :base_name,
+                                           :number, :base_branch, :head_sha, :overrides
+
+            # @param [CachedPullRequest] cache
+            # @param [Github::PullRequest] pull_request
+            # @return [Boolean]
+            def same?(cache, pull_request)
```

Comment:
Would IMO make more sense to have an `caches_pull_request?` method on `CachedPullRequest`. For info, one can add methods to a Struct with a block:

~~~
CachedPullRequest = Struct.new :foo, :bar do
  def caches_pull_request?
  end
end
~~~

If you do want to stick with the method on `PullRequestCache`, I think you need to find a more descriptive name.

Reply:
> Would IMO make more sense to have an caches_pull_request? method on CachedPullRequest

That's a good idea but I think it is still good to have `same?` (or some renamed version of it) in `PullRequestCache` to avoid adding the same pull request twice

Reply:
Don't get it ... it's the same functionality, just placed differently (i.e. can be called all the same)

Reply:
I don't think we are talking about the same thing. 

- `#same?`  is "are these two pull requests the same?" (same owner, same repository, same number). 
- `#cached?`  is "do these pull requests (which are the same) have equal metadata/info?" (equal base branch, equal head commit and equal comment body)?

`#same?` is used by `PullRequestCache` to prevent the same `Github::PullRequest` from being added twice. `#cached?` is used to find out whether the `Github::PullRequest` have been updated w.r.t to `PullRequestCache`'s internal `CachedPullRequest` version of it.

Reply:
Is `cached?` actually `changed?` ?

`#same?`'s first argument is a CachePullRequest, so you're checking whether the pull request is already present in cache (may need cache invalidation, but this is what it is checking).

I am still very confused reading the code ... Either there's something I'm still missing (hence confusion ...), or I do believe it would be a lot readable if `#same?` would be called `#from_pull_request?` on `CachedPullRequest`.

Reply:
I made the confusion with the argument type, indeed, both methods compare CachedPullRequest to PullRequest even though they do different things. I will move both to the struct